File: languages/elixir/exercises/concept/regular_expressions/lib/date_parser.ex
defmodule DateParser do
  def month(), do: "\\d{1,2}"
  def day(), do: "\\d{1,2}"
  def year(), do: "\\d{4}"

  @days ~w[Sunday Monday Tuesday Wednesday Thursday Friday Saturday]
  def day_names() do
    "(#{Enum.join(@days, "|")})"
  end

  @months ~w[January February March April May June July August September October November December]
  def month_names() do
    "(#{Enum.join(@months, "|")})"
  end

  def capture_month(), do: "(?<month>#{month()})"
  def capture_day(), do: "(?<day>#{day()})"
  def capture_year(), do: "(?<year>#{year()})"

  def capture_month_names(), do: "(?<month_names>#{month_names()})"
  def capture_day_names(), do: "(?<day_names>#{day_names()})"

  def capture_numeric_date(), do: "#{capture_day()}/#{capture_month()}/#{capture_year()}"
  def capture_month_name_date(), do: "#{capture_month_names()} #{capture_day()}, #{capture_year()}"
  def capture_day_month_name_date(), do: "#{capture_day_names()}, #{capture_month_name_year()}"

  def match_numeric_date(), do: ~r'^#{capture_numeric_date()}$'
  def match_month_name_date(), do: ~r'^#{capture_month_name_date()}$'
  def match_day_month_name_date(), do: ~r'^#{capture_day_month_name_date()}$'
end

File: languages/clojure/exercises/concept/date-parser/src/exercism/date_parser.clj
(ns exercism.date-parser)

(def day "\\d{1,2}")
(def month "\\d{1,2}")
(def year "\\d{4}")

(def days "Sunday|Monday|Tuesday|Wednesday|Thursday|Friday|Saturday")

(defn day-names [s]
  (re-matches (re-pattern days) s))

(def months "January|February|March|April|May|June|July|August|September|October|November|December")

(defn month-names [s]
  (re-matches (re-pattern months) s))

(defn capture-month [s]
  (let [matcher (re-matcher (re-pattern (str "(?<month>" month ")")) s)]
    (when (.matches matcher)
      {:month (.group matcher "month")})))

(defn capture-day [s]
  (let [matcher (re-matcher (re-pattern (str "(?<day>" day ")")) s)]
    (when (.matches matcher)
      {:day (.group matcher "day")})))

(defn capture-year [s]
  (let [matcher (re-matcher (re-pattern (str "(?<year>" year ")")) s)]
    (when (.matches matcher)
      {:year (.group matcher "year")})))

(defn capture-month-name [s]
  {:month-name   (first (re-find (re-pattern (str "(?<month>" months ")")) s))})

(defn capture-day-name [s]
  {:day-name (first (re-find (re-pattern (str "(?<day>" days ")")) s))})

(defn capture-numeric-date [s]
  (let [matcher (re-matcher (re-pattern (str "(?<day>" day ")/(?<month>" month ")/(?<year>" year ")")) s)]
    (when (.matches matcher)
      {:day (.group matcher "day")
       :month (.group matcher "month")
       :year (.group matcher "year")})))

(defn capture-month-name-date [s]
  (let [matcher (re-matcher (re-pattern (str "(?<month>" months ") (?<day>" day "), (?<year>" year ")")) s)]
    (when (.matches matcher)
      {:month-name (.group matcher "month")
       :day (.group matcher "day")
       :year (.group matcher "year")})))

(defn capture-day-month-name-date [s]
  (let [matcher (re-matcher (re-pattern (str "(?<dayname>" days "), (?<month>" months ") (?<day>" day "), (?<year>" year ")")) s)]
    (when (.matches matcher)
      {:day-name (.group matcher "dayname")
       :month-name (.group matcher "month")
       :day (.group matcher "day")
       :year (.group matcher "year")})))
